cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/AbstractApplicationDetailFetcher.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.metadata;
+
+import io.cdap.cdap.common.NotFoundException;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.id.ApplicationReference;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * Abstract implementation for {@link ApplicationDetailFetcher} for common functionalities.
+ */
+public abstract class AbstractApplicationDetailFetcher implements ApplicationDetailFetcher {
+
+  /**
+   * Current implementation is suboptimal as it fetches each application individually either in
+   * separate db transaction or separate api call.
+   * TODO(samik) Add batch fetch by id in DefaultStore and expose via API
```

Comment:
I don't think we should add this method until we actually have a batch method. Especially since I'm not sure we actually want it either, it seems like a pre-mature optimization. For example, if we're holding the transaction open while scanning, it can potentially cause a lot of contention on the table, especially if we're doing expensive operations while scanning.

I think it's fine to loop through with individual fetches. We don't currently have a requirement to read them all in the same transaction, the transactions will be much shorter, and it shouldn't be that much slower, especially compared to other I/O that we're doing.

Reply:
I agree. This iterator actually was not giving any extra help and will only be used in push operation anyway. I will remove the iterator and directly use the single get.

Reply:
Removed changes to `ApplicationFetcher`